project/src/main/ui/chat/career-cutscene-library.gd

```diff
@@ -0,0 +1,309 @@
+		for child in children:
+			var child_key := _child_key(chat_key_roots, chat_key, child)
+			chat_key_queue.push_front(child_key)
+		
+		postorder_chat_keys.push_front(chat_key)
+	return postorder_chat_keys
+
+
+## Filters the list of potential chat key pairs, excluding the specified chat keys.
+##
+## The returned list includes the earliest numeric keys in each branch (0, 1, 2 are always played in order) as well as
+## ALL alphabetic keys in each branch (a, b, c are ordered arbitrarily)
+##
+## Parameters:
+## 	'chat_key_roots': An array of string chat key roots like 'chat/career/marsh' which correspond to a group of
+## 		cutscenes to choose between.
+##
+## 	'exhausted_chat_keys': A set of chat keys for cutscenes the player has already seen. This includes the leaf
+## 		nodes for cutscenes which have been seen, and branch nodes where all child cutscenes have been seen.
+##
+## Returns:
+## 	A filtered list of dictionary entries. Each dictionary entry has a 'preroll' entry and/or a 'postroll' entry,
+## 	defining chat keys for cutscenes which play before or after a level. The list is filtered to exclude cutscenes
+## 	the player has either already seen, or should not see until later.
+func _filter_chat_key_pairs(chat_key_roots: Array, exhausted_chat_keys: Dictionary) -> Array:
+	var potential_chat_key_pairs := []
+	var chat_key_queue := chat_key_roots.duplicate()
+	var excluded_chat_keys := exhausted_chat_keys.duplicate()
+	while chat_key_queue:
+		var chat_key: String = chat_key_queue.pop_front()
+		if excluded_chat_keys.has(chat_key):
+			# don't include these chat keys; the player's already seen these cutscenes
+			continue
+		
+		excluded_chat_keys[chat_key] = true
+		
+		var children: Array = _preroll_tree.get(chat_key, [])
+		if not children:
+			# leaf node; enqueue it
+			potential_chat_key_pairs.append(_chat_key_pairs_by_preroll[chat_key])
+		else:
+			# branch node; enqueue its children
+			var min_numeric_child := ""
+			for child in children:
+				# enqueue letter key
+				var child_key := _child_key(chat_key_roots, chat_key, child)
+				if excluded_chat_keys.has(child_key):
+					# don't include these chat keys; the player's already seen these cutscenes
+					pass
+				elif child.is_valid_integer():
+					if not min_numeric_child or int(child) < int(min_numeric_child):
+						# track lowest-numbered numeric key
+						min_numeric_child = child
+				else:
+					if not excluded_chat_keys.has(child_key):
+						chat_key_queue.push_back(child_key)
+				
+			# enqueue lowest-numbered numeric key
+			if min_numeric_child:
+				var child_key := _child_key(chat_key_roots, chat_key, min_numeric_child)
+				if not excluded_chat_keys.has(child_key):
+					chat_key_queue.push_front(child_key)
+	
+	return potential_chat_key_pairs
+
+
+func _child_key(chat_key_roots: Array, chat_key: String, child_suffix: String) -> String:
+	return chat_key + ("/" if chat_key in chat_key_roots else "_") + child_suffix
```

Comment:
This is a weird function; comment it

Reply:
Added comment, `Returns a child chat key by combining the specified chat key ...`